polar plots do not properly handle units

jrevans opened the pull request:
PolarAxes was not properly handling unitized data

This addresses an issue in #4897.

tacaswell commented on lib/matplotlib/projections/polar.py:
```diff
@@ -521,9 +523,17 @@ def set_thetagrids(self, angles, labels=None, frac=None, fmt=None,
+           angles = np.asarray(angles, np.float_)
+           self.set_xticks(angles * (np.pi / 180.0))
+        else:
+           # The unit converters are defined to return radians
```
Silly question, where is this documented/implemented?

QuLogic commented on lib/matplotlib/projections/polar.py:
```diff
@@ -521,9 +523,17 @@ def set_thetagrids(self, angles, labels=None, frac=None, fmt=None,
-        self.set_xticks(angles * (np.pi / 180.0))
+        if isinstance( angles[0], float ):
+           # If the data is floats it is assumed to be degrees (as per the
+           # docsting for this method)
```
docstring

QuLogic commented on lib/matplotlib/projections/polar.py:
```diff
@@ -521,9 +523,17 @@ def set_thetagrids(self, angles, labels=None, frac=None, fmt=None,
+           self.set_xticks(angles * (np.pi / 180.0))
+        else:
+           # The unit converters are defined to return radians
+           angles = self.convert_yunits(angles)
```
Shouldn't this be `x`?

tacaswell replied:
I don't understand you `x`, `y` related comments.  These are internal variables and named to be consistent with the input.  If we should have essentially wrapper functions for x/y ticks on polar axes is a ship that is sailed.

QuLogic replied:
Matplotlib's polar plots are `(theta, r)`; the `convert_[xy]units` calls assume `(r, theta)`.

WeatherGod replied:
ah, I think I know what you are getting at. Yes, radius is the considered the "y-axis", while theta is considered the "x-axis" in the sense that that is what you use for calling tickers and labels (so, ax.set_xlabels() for setting the theta labels). However, it appears that `convert_xunits()` has traditionally been used for radius and `convert_yunits()` have been traditionally used for theta. I don't know if that is a bug or not. Perhaps this never worked correctly?

QuLogic replied:
That is... odd. Pretty much everything works fine if you want to forgo the convenience wrappers and use `(x, y)` as `(theta, r)`. It's just strange that unit conversion would do the opposite.

And this sounds buggy. If you call `set_rticks`, it directly calls `set_yticks`, so the converter will be set on the y-axis. Then if you call `set_thetagrids`, the y-axis is used, so that same converter used for radius will be applied to the angles.

QuLogic commented on lib/matplotlib/projections/polar.py:
```diff
@@ -566,6 +576,10 @@ def set_rgrids(self, radii, labels=None, angle=None, fmt=None,
 
+        # Handle unitized data
+        if angle is not None:
+            angle = self.convert_yunits(angle)
```
And `x` here as well.

QuLogic commented on lib/matplotlib/projections/polar.py:
```diff
@@ -591,6 +605,14 @@ def format_coord(self, theta, r):
         characters.
         """
+        # Strip away any units
+        r = self.convert_xunits(r)
```
Should be `y`.

QuLogic commented on lib/matplotlib/projections/polar.py:
```diff
@@ -591,6 +605,14 @@ def format_coord(self, theta, r):
+        # Strip away any units
+        r = self.convert_xunits(r)
+        if not isinstance( theta, float ):
+            theta = self.convert_yunits(theta)
```
And `x` here.

QuLogic commented:
Side note: there's no reason to prefix your commit messages with your initials; that's why you set your author info before committing anything.

QuLogic commented on lib/matplotlib/projections/polar.py:
```diff
@@ -399,6 +399,8 @@ def set_theta_offset(self, offset):
         Set the offset for the location of 0 in radians.
         """
+        # Make sure to strip away units
+        offset = self.convert_yunits(offset)
```
And `x` here too.

WeatherGod commented on lib/matplotlib/projections/polar.py:
```diff
@@ -521,9 +523,17 @@ def set_thetagrids(self, angles, labels=None, frac=None, fmt=None,
-        angles = self.convert_yunits(angles)
-        angles = np.asarray(angles, np.float_)
-        self.set_xticks(angles * (np.pi / 180.0))
+        if isinstance( angles[0], float ):
```
Why do we need this if-branch? Is `convert_yunits()` not returning radians? If not, is that a regression?

WeatherGod commented on lib/matplotlib/projections/polar.py:
```diff
@@ -591,6 +605,14 @@ def format_coord(self, theta, r):
+            theta = self.convert_yunits(theta)
+            # Stripping away units makes theta be in radians.
+            # convert to degrees
+            theta = theta * (180.0 / np.pi)
```
Maybe I am wrong, but this seems incorrect? The subsequent code already divides theta by pi and later multiplies by 180. In fact, the formatting string says that it is displaying radians first, then degrees second.

QuLogic replied:
I think you're right. Even though this function calls the arguments `theta`/`r`, it's called by matplotlib, so it's (sorta) `x`/`y`, meaning "`theta`" is in radians.

tacaswell commented:
Can you modify or extend https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_axes.py#L384 to catch this issue?

jrevans commented:
When moving my fixes over to git, I had accidentally transposed which axis was used for the unit conversion.  I have fixed that.

I have also added some more test cases to test for this.  I am not sure what the process is for generating the baseline images or even running the test harness anymore.  It has changed dramatically from when I first wrote it up so many years ago.  What is the process for this now?

tacaswell commented:
See
http://matplotlib.org/devel/testing.html#writing-an-image-comparison-test
for the current directions

On Mon, Aug 17, 2015, 3:54 PM James Evans <notifications@github.com> wrote:

> When moving my fixes over to git, I had accidentally transposed which axis
> was used for the unit conversion. I have fixed that.
>
> I have also added some more test cases to test for this. I am not sure
> what the process is for generating the baseline images or even running the
> test harness anymore. It has changed dramatically from when I first wrote
> it up so many years ago. What is the process for this now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4905#issuecomment-131942453>
> .
>

WeatherGod commented:
Essentially, just run the test directly (after installing the latest
changess). So, from the root source directory: `python
lib/matplotlib/tests/test_foobar.py`). It'll run and error on the new test,
but it will produce artifacts like
"result_images/test_foobar/this_new_test_image.png". Verify the new
artifacts are right and then copy them over to
"lib/matplotlib/tests/baseline_images/test_foobar/". Rerun the test to make
sure all is well, and then add and commit those images.

On Mon, Aug 17, 2015 at 3:54 PM, James Evans <notifications@github.com>
wrote:

> When moving my fixes over to git, I had accidentally transposed which axis
> was used for the unit conversion. I have fixed that.
>
> I have also added some more test cases to test for this. I am not sure
> what the process is for generating the baseline images or even running the
> test harness anymore. It has changed dramatically from when I first wrote
> it up so many years ago. What is the process for this now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4905#issuecomment-131942453>
> .
>

tacaswell commented:
ping @jrevans Any update on this?

mdboom commented on lib/matplotlib/projections/polar.py:
```diff
@@ -591,11 +605,21 @@ def format_coord(self, theta, r):
         # \u00b0: degree symbol
-        return '\u03b8=%0.3f\u03c0 (%0.3f\u00b0), r=%0.3f' % (theta, theta * 180.0, r)
+        s = '\u03b8=%0.3f\u03c0 (%0.3f\u00b0), r=%0.3f' % (theta, theta * 180.0, r)
+        s = s.encode( 'utf-8' )
```
This shouldn't be necessary and probably indicates a bug somewhere else.  Returning unicode from this function is actually what we *want*.  Can you describe which backend/platform you were using when you can across this problem?

mdboom commented:
@jrevans: Any progress on addressing the questions above?  We may have to punt on this for 1.5.0.

tacaswell commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -411,6 +411,107 @@ def test_polar_units():
     assert_true(isinstance(plt.gca().get_xaxis().get_major_formatter(), units.UnitDblFormatter))
 
 
+def test_polar_format_coord():
```
This needs a `@cleanup` decorator so global state does not leak.

mdboom commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -411,6 +411,107 @@ def test_polar_units():
+    ax = plt.subplot(111, polar=True)
+
+    # Make sure this doesn't throw any exceptions
+    ax.format_coord( 30.0*deg, 4.0*km )
```
Can you PEP8 the whitespace here?

mdboom commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -411,6 +411,107 @@ def test_polar_units():
+    ax.plot( x1, y1 )
+
+
+@image_comparison(baseline_images=['polar_teta_offset'])
```
`teta` -> `theta`

mdboom commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -411,6 +411,107 @@ def test_polar_units():
+    ax.format_coord( 30.0*deg, 4.0*km )
+
+
+@image_comparison(baseline_images=['polar_rgrids'])
```
As these tests add new image comparison tests, you'll also need to commit the new baseline images.

jklymak commented:
Closing this as not likely to be resurrected, but feel free to request a re-open.
